pkg/openfeature/client.go

```diff
@@ -65,126 +70,198 @@ type EvaluationDetails struct {
+) (EvaluationDetails, error) {
+	var err error
+	hookCtx := HookContext{
+		flagKey:           flag,
+		flagType:          flagType,
+		defaultValue:      defaultValue,
+		clientMetadata:    c.metadata,
+		providerMetadata:  api.provider.Metadata(),
+		evaluationContext: evalCtx,
 	}
-	return EvaluationDetails{
-		FlagKey:          flag,
-		FlagType:         Boolean,
+	evalDetails := EvaluationDetails{
+		FlagKey:  flag,
+		FlagType: flagType,
 		ResolutionDetail: ResolutionDetail{
-			Value:     value,
-			ErrorCode: resolution.ErrorCode,
-			Reason:    resolution.Reason,
-			Variant:   resolution.Variant,
+			Value: defaultValue,
 		},
-	}, err
+	}
 
-}
+	apiClientInvocationHooks := append(append(api.hooks, c.hooks...), options.hooks...) // API, Client, Invocation
+	invocationClientApiHooks := append(append(options.hooks, c.hooks...), api.hooks...) // Invocation, Client, API
+	defer func() {
+		c.finallyHooks(hookCtx, invocationClientApiHooks, options)
+	}()
 
-// GetStringValueDetails return string evaluation for flag
-func (c Client) GetStringValueDetails(flag string, defaultValue string, evalCtx EvaluationContext, options ...EvaluationOption) (EvaluationDetails, error) {
-	resolution := api.provider.GetStringEvaluation(flag, defaultValue, evalCtx, options...)
-	value := resolution.Value
-	err := resolution.Error()
+	evalCtx, err = c.beforeHooks(hookCtx, apiClientInvocationHooks, evalCtx, options)
+	hookCtx.evaluationContext = evalCtx
 	if err != nil {
-		value = defaultValue
+		err = fmt.Errorf("execute before hook: %w", err)
+		c.errorHooks(hookCtx, invocationClientApiHooks, err, options)
+		return evalDetails, err
 	}
-	return EvaluationDetails{
-		FlagKey:          flag,
-		FlagType:         String,
-		ResolutionDetail: ResolutionDetail{
-			Value:     value,
-			ErrorCode: resolution.ErrorCode,
-			Reason:    resolution.Reason,
-			Variant:   resolution.Variant,
-		},
-	}, err
-}
 
-// GetNumberValueDetails return number evaluation for flag
-func (c Client) GetNumberValueDetails(flag string, defaultValue float64, evalCtx EvaluationContext, options ...EvaluationOption) (EvaluationDetails, error) {
-	resolution := api.provider.GetNumberEvaluation(flag, defaultValue, evalCtx, options...)
-	value := resolution.Value
-	err := resolution.Error()
+	var resolution ResolutionDetail
+	switch flagType {
+	case Object:
+		resolution = api.provider.GetObjectEvaluation(flag, defaultValue, evalCtx, options)
+	case Boolean:
+		defValue := defaultValue.(bool)
+		res := api.provider.GetBooleanEvaluation(flag, defValue, evalCtx, options)
+		resolution = res.ResolutionDetail
+	case String:
+		defValue := defaultValue.(string)
+		res := api.provider.GetStringEvaluation(flag, defValue, evalCtx, options)
+		resolution = res.ResolutionDetail
+	case Number:
+		defValue := defaultValue.(float64)
+		res := api.provider.GetNumberEvaluation(flag, defValue, evalCtx, options)
+		resolution = res.ResolutionDetail
```

Comment:
Because the caller already "knows" the type, Instead of a switch case here, I _think_ you can use a reference to the appropriate provider method passed from the caller. I did something similar in JS: https://github.com/open-feature/node-sdk/blob/66ed1b1a56e5fa1137ff66198ac245c5c8496347/src/client.ts#L146

The only thing that might be a challenge is defining the receiver function type generally and accurately without generics since they have different params. If you have to use `interface{}` for the function, it might not be worth it.

Reply:
Without generics the only way that approach would work is if the provider functions all had the exact same signature. Using `interface{}` as a match all wouldn't work in this sense, the provider functions themselves would have to define their parameters as `interface{}`

Reply:
Alas!